txtorcon/torconfig.py

```diff
@@ -1129,13 +1129,24 @@ def save(self):
             return defer.succeed(self)
 
         args = []
+        directories = []
+        hidden_services = []
         for (key, value) in self.unsaved.items():
             if key == 'HiddenServices':
-                self.config['HiddenServices'] = value
                 for hs in value:
                     for (k, v) in hs.config_attributes():
-                        args.append(k)
-                        args.append(v)
+                        if k == 'HiddenServiceDir':
+                            if v not in directories:
+                                directories.append(v)
+                                hidden_services.append(hs)
+                                args.append(k)
+                                args.append(v)
+                            else:
+                                raise RuntimeError("Trying to add hidden service with same HiddenServiceDir: %s" % v)
+                        else:
+                            args.append(k)
+                            args.append(v)
+                self.config['HiddenServices'] = hidden_services
```

Comment:
Thanks for pointing that out. We do not need to remake the list as if there are redundant hs-dirs, we will get an error anyways and it wont get saved.
I will update the code in the next commit